### PR TITLE
GH-34909: [C++] Avoid mean overflow on large integer inputs

### LINKED ISSUE #34909: [C++] mean overflows if numeric sum is larger than int64 max
### Describe the bug, including details regarding any error messages, version, and platform.

```
>>> arr = pa.array([-1303487490025821099, -8371390547526583103, -2572374159461887095], type=pa.int64())
>>> pa.compute.mean(arr)
<pyarrow.DoubleScalar: 2.0664972922317535e+18>
>>>
```

When using numpy:

```
>>> arr.to_numpy().mean()
-4.0824173990047636e+18
>>>
```

Seems like overflow case is not handled correctly when using arrow array.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by js8544**
### Rationale for this change

The `mean` aggregate function would overflow if the input array's sum is larger than int64_max.

### What changes are included in this PR?

Store intermediate sum in double instead of int64, so that it won't overflow.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.

* Closes: #34909

### CODE REVIEW

**Comment by Light-City** [comment]
In the current implementation, the overflow problem is not considered

**Comment by Light-City** [comment]
> In the current implementation, the overflow problem is not considered

just like:

```
  func = std::make_shared<ScalarAggregateFunction>("sum", Arity::Unary(), sum_doc,
                                                   &default_scalar_aggregate_options);
  AddArrayScalarAggKernels(SumInit, {boolean()}, uint64(), func.get());
  AddAggKernel(KernelSignature::Make({Type::DECIMAL128}, FirstType), SumInit, func.get(),
               SimdLevel::NONE);
  AddAggKernel(KernelSignature::Make({Type::DECIMAL256}, FirstType), SumInit, func.get(),
               SimdLevel::NONE);
  AddArrayScalarAggKernels(SumInit, SignedIntTypes(), int64(), func.get());
```

sum(int64)-> output also is int64(), will cause overflow.

**Comment by js8544** [comment]
What you said is correct. However, this PR is intended to deal with the `mean` function, not `sum`. I'll add a `sum_checked` function after this PR is merged.

File: cpp/src/arrow/compute/kernels/aggregate_basic_internal.h [hunk 1]
```diff
@@ -196,20 +203,32 @@ struct MeanImpl : public SumImpl<ArrowType, SimdLevel> {
...
     return Status::OK();
   }
+
+  Status Finalize(KernelContext*, Datum* out) override { return FinalizeImpl(out); }
+};
+
+template <typename ArrowType, SimdLevel::type SimdLevel>
+struct MeanImpl<ArrowType, SimdLevel,
+                std::enable_if_t<!is_decimal_type<ArrowType>::value>>
+    : public SumImpl<ArrowType, SimdLevel, DoubleType> {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_basic_internal.h:213]: Can you add a comment above explaining that you're overriding the sum result type and why?

**Review by pitrou — CHANGES_REQUESTED** [review]
One more comment. Also, can you rebase/merge from main?

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you @js8544 !

**Pull request merged**